### DISCUSSION TITLE
Inherent Usability Issue outside of OOP

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#4932] [Ideas]
I've programmed some small Flix projects and felt a certain friction when writing code. My default language for a long time has been Scala, so that is where I'm coming from. I think I've nailed a big part of it down to the basic operation of calling a function:

### OOP workflow
The scenario is this: I want to do something to a value of type `A`. In OOP, I would type `a.`, and through the IDE I can see all of the class members. This feature helps me in two parts.
1. I can find my function even through typos, and I don't have to spell it out
2. I can see what I can do with this object in an exploratory way, not necessarily looking for a name!

Part 1 works perfectly fine in FP, and Flix is getting better at it, but I really miss part 2. Another more subtle benefit is that I don't have to bother with namespacing in the OOP world; it's the class member!

### FP workflow
In FP/Flix I would have my thing `a` and if I don't know have a name, then I will try `A.` and see what comes up. Well:
1. Before even getting to this point, I have to know if the namespace is `MyProject/A` or `ThirdParty/Util/A` or whatever
2. I can see every function in the namespace, including irrelevant functions like `fromB` that cannot operate on a value of `A`.
3. I am completely lost if what I'm looking for is not in `A` (which is also true in OOP! fx. in Java, your goal might be in `Collections` instead of on the object)

### What to do about it
I realized that what I like in OOP is not really what I want, just a decent approximation; I want to find things I can do to `A`! I think theorems provers do this quite well, even though the ergonomics are not always great.

I want to be able to query `List[String]` and see all the callable functions that take such an argument. Of course, there are many ways to expand this with more complex queries, but just the baseline would be fantastic, and I don't think autocomplete can realistically fulfill this purpose.

**Comment by magnus-madsen** [comment]
If you could come up with some auto-completion strategy we could try to implement it. 

Perhaps something on `a |> `?

**Reply by JonathanStarup** [reply]
# Hole-directed type search
I'm not sure about autocomplete, can we know the type of the cursor position while typing? It could maybe work with a hole thing like `?helpme`, and if it compiles, then you have a function type that you can search for. We could even add a vscode quick action to fill it in.

Pros:
* Minimum infrastructure needed
* Can help with suggestions, even local variables
Cons:
* Needs a correct program
* Strict query system. If I'm thinking about how to convert `Int64` to `Int32`, then I have to construct such a hole.

I definitely think the above would be nice; I image typing my function with a couple `?helpme` and then gradually the compiler helps me complete them. This seems like nice top-down programming.

# External type search
I think I would also like a way to search for things without having to construct a hole. My ideal would be something like `#...` search in vscode but that doesn't seem possible from what I can see, so the second best thing is the REPL. Essentially expanding on the `:doc` command. (could also be its own thing). The REPL has the advantage that has the last working program (excluding errors on startup).

```
flix>  :doc Int32 -> String
- Int32.toString(x: Int32): String
- Local/Ns.swirl(x: Int32): String
```
This could be as simple as a concrete type like the above, to more complicated patterns like
* `(Int32, ...) -> String`: something that takes additional arguments together with `Int32`
* `Int32 -> _` wildcards
* `Int32 -> String in Local/*`:  namespace filtered

---

I know this might be a bit niche, and I think `api.flix.dev` helps a lot, but I really like asking my runtime since I don't have to know some URL and there is no version confusion. I think exploration via REPL can be really valuable. It's not a fully thought out idea but I hope it seems sensible

**Reply by JonathanStarup** [reply]
It seems to me that the two systems above can be implemented largely with the same compiler support

**Comment by magnus-madsen** [comment]
Quick comment:

I think @Daniel-Anker-Hermansen will try to support auto-complete on `???` and `?helpme`.

In addition, I will try to add the ability to write `myVar?` as a hole.

**Reply by JonathanStarup** [reply]
* I don't know about the technical aspect, but I feel that this should be a quick action instead of autocomplete since you don't necessarily want to replace the `???`
* maybe just `?help` instead of `?helpme`

**Comment by magnus-madsen** [comment]
I think in addition that @Daniel-Anker-Hermansen can add some special casing for `|> ???`.
